### PR TITLE
Release v4.0.0-rc

### PR DESCRIPTION
**Pull request by smiasojed**
### Added
- Add schema generation and verification - [#1404](https://github.com/paritytech/cargo-contract/pull/1404)
- Compare `Environment` types against the node - [#1377](https://github.com/paritytech/cargo-contract/pull/1377)
- Detect `INK_STATIC_BUFFER_SIZE` env var - [#1310](https://github.com/paritytech/cargo-contract/pull/1310)
- Add `verify` command - [#1306](https://github.com/paritytech/cargo-contract/pull/1306)
- Add `--binary` flag for `info` command - [#1311](https://github.com/paritytech/cargo-contract/pull/1311/)
- Add `--all` flag for `info` command - [#1319](https://github.com/paritytech/cargo-contract/pull/1319)
- Add contract language detection feature for `info` command - [#1329](https://github.com/paritytech/cargo-contract/pull/1329)
- Add warning message when using incompatible contract's ink! version - [#1334](https://github.com/paritytech/cargo-contract/pull/1334)
- Add workspace support -[#1358](https://github.com/paritytech/cargo-contract/pull/1358)
- Add `Storage Total Deposit` to `info` command output - [#1347](https://github.com/paritytech/cargo-contract/pull/1347)
- Add dynamic types support - [#1399](https://github.com/paritytech/cargo-contract/pull/1399)
- Basic storage inspection command - [#1395](https://github.com/paritytech/cargo-contract/pull/1395)

### Changed
- Bump `subxt` to `0.32.0` - [#1352](https://github.com/paritytech/cargo-contract/pull/1352)
- Remove check for compatible `scale` and `scale-info` versions - [#1370](https://github.com/paritytech/cargo-contract/pull/1370)
- Bump `ink` to `5.0.0-rc` - [#1415](https://github.com/paritytech/cargo-contract/pull/1415)

### Fixed
- Do not allow to execute calls on immutable contract messages - [#1397](https://github.com/paritytech/cargo-contract/pull/1397)
- Improve JSON Output for Upload and Remove Commands - [#1389](https://github.com/paritytech/cargo-contract/pull/1389)
- Fix for a Url to String conversion in `info` command - [#1330](https://github.com/paritytech/cargo-contract/pull/1330)

### CODE REVIEW

File: crates/extrinsics/src/env_check.rs [hunk 1]
```diff
@@ -287,7 +287,7 @@ mod tests {
                         vec!["i32"].into_iter().map(AsRef::as_ref),
                     ))
                     .done()])
-                .returns(ReturnTypeSpec::new(None))
+                .returns(ReturnTypeSpec::new(TypeSpec::default()))
```
Comment (SkymanOne) [inline crates/extrinsics/src/env_check.rs:290]: The default value is: 
```rust
ReturnTypeSpec::new(
    TypeSpec::with_name_str::<ink_primitives::ConstructorResult<()>>("ink_primitives::ConstructorResult")
)
```

Not sure if `ink_primitives` is imported.

Comment (smiasojed) [reply crates/extrinsics/src/env_check.rs:290]: It is not imported; that is why I used the default.

File: crates/transcode/src/lib.rs [hunk 2]
```diff
@@ -400,22 +400,16 @@ impl ContractMessageTranscoder {
...
         let msg_spec = self.find_message_spec(name).ok_or_else(|| {
             anyhow::anyhow!("Failed to find message spec with name '{}'", name)
         })?;
-        if let Some(return_ty) = msg_spec.return_type().opt_type() {
-            self.decode(return_ty.ty().id, data)
-        } else {
-            Ok(Value::Unit)
-        }
+        let return_ty = msg_spec.return_type().ret_type();
+        self.decode(return_ty.ty().id, data)
```
Comment (ascjones) [inline crates/transcode/src/lib.rs:412]: Need to double check that this still works by doing a `cargo contract call` on a message with a return value

**Review by ascjones — APPROVED** [review]
LGTM

**Review by cmichi — APPROVED** [review]

**Review by SkymanOne — CHANGES_REQUESTED** [review]
Need to update readme to include `cargo-analyzer` in the release instructions.

**Pull request merged**